### PR TITLE
ast: Record docstrings

### LINKED ISSUE #319: AST: Support parsing docstrings
We should parse docstrings and make them available in the AST.

### PR DESCRIPTION
**Pull request by abhinav**
This adds support for reading docstrings from Thrift IDLs.

We currently parse docstrings specified for,

- [x] Parse docstrings for all entities.
  - [x] Top-level definitions
  - [x] Enum items
  - [x] Struct fields
  - [x] Service functions
  - [x] Function parameters
  - [x] Function exceptions
- [x] Clean up docstrings. We need to remove the leading spaces and `*` from
      each line in the docstring based on the indentation, etc. That is,
      
          /**
           * foo
           * bar
           */
           
    Should be interpreted as,
      
          foo
          bar
      
    Not,
      
          * foo
          * bar
- [x] Tests

Follow up tasks:

- [ ] Include docstrings in generated code

Resolves #319

### CODE REVIEW

**Comment by abhinav** [comment]
This diff is ready to review.

File: idl/internal/docstring.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+// CleanupDocstring takes a docstring in the form,
+//
+//  /**
+//   * foo bar
+//   */
+//
+// And returns,
+//
+//  foo bar
+func CleanupDocstring(s string) string {
```
Comment (breerly) [inline idl/internal/docstring.go]: Cleanup feels weird here, suggest `ParseDocstring` instead.

Comment (prashantv) [reply idl/internal/docstring.go]: Or maybe `FormatDocstring`?

File: idl/internal/docstring.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	lines := strings.Split(s, "\n")
+	dedent(lines, true /* skipFirstIfUnindented */)
+
+	switch len(lines) {
+	case 0:
+		return s
```
Comment (prashantv) [inline idl/internal/docstring.go]: don't think this case can happen,
https://play.golang.org/p/fswgcibGYQ

File: idl/internal/docstring.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	case 1:
+		// /** foo */
+		s := lines[0]
+		s = strings.TrimPrefix(s, _docstringOpen)
+		s = strings.TrimSuffix(s, _docstringClose)
+		return strings.TrimSpace(s)
+	}
+
+	if len(lines) > 0 {
```
Comment (prashantv) [inline idl/internal/docstring.go]: We must have `len(lines) > 0` if we got to this point

File: idl/internal/docstring.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		lines[0] = strings.TrimPrefix(lines[0], _docstringOpen)
+		lines[len(lines)-1] = strings.TrimSuffix(lines[len(lines)-1], _docstringClose)
+	}
```
Comment (prashantv) [inline idl/internal/docstring.go]: this is basically the same logic that the `case 1` also has, trim the `_docstringOpen` and `_docstringClose` from the first and last line. Maybe the logic should look like:
```go
lines := strings.Split(...)
last := len(lines)-1

// Trim the /** and */ from the first and last lines.
lines[0] = strings.TrimPrefix(lines[0], _docstringOpen)
lines[last] = strings.TrimSuffix(lines[last], _docstringClose)

if len(lines) == 1 {
  return strings.TrimSpace(lines[0])
}

lines = dropLeadingEmptyLines(lines)
[...]
```

File: idl/internal/docstring_test.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		},
+		{
+			give: []string{
+				"/**",
+				"	    * foo does stuff",
+				"	    */",
+			},
+			want: []string{"foo does stuff"},
+		},
+	}
```
Comment (prashantv) [inline idl/internal/docstring_test.go]: can you add a test case where the lines inside are missing the ` * `? e.g.,
```
give : []string{
  "/**",
  "no prefix",
  "for lines",
  "*/",
},
```

and maybe another case where we have a mix of lines with the prefix and without

File: idl/internal/docstring.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		lines = lines[:i]
+	}
+	return lines
+}
+
+// skipFirstIfUnindented tells dedent to not consider the first line for the
+// prefix if it's not already indented. This is needed because our docstringcs
+// will start at "/**" without any leading whitespace.
+func dedent(lines []string, skipFirstIfUnindented bool) {
```
Comment (prashantv) [inline idl/internal/docstring.go]: don't really understand what this function is supposed to do, can you put a short description in the comment above

File: idl/internal/docstring.go (hunk 1 above)
Comment (abhinav) [reply idl/internal/docstring.go]: I like Parse more than Format here. This isn't formatting a docstring to be printed but parsing what the actual text from a docstring.

File: idl/internal/docstring.go (hunk 2 above)
Comment (abhinav) [reply idl/internal/docstring.go]: That's probably true but I would like to keep the check in for the sake of defensive programming.

File: idl/internal/docstring.go (hunk 3 above)
Comment (abhinav) [reply idl/internal/docstring.go]: Good catch with this and the other comment. A bunch of code got moved around but I forgot to remove the checks that were now redundant.

File: idl/internal/docstring.go (hunk 6 above)
Comment (abhinav) [reply idl/internal/docstring.go]: Added. My original thought was that dedent would be self-explanatory as the reverse of indent except this function became really complicated.

**Comment by abhinav** [comment]
Updated. The goal is to make a best-effort attempt of extracting the paragraph inside the docstring.

In,

```
struct X {
  /**
   * foo
   * bar
   */
  1: string foo
}
```

The string we get for the docstring of the field is,

```
/**
   * foo
   * bar
   */
```

We need to extract everything on the right side of the `*`s, preserve
indentation, and ideally, handle cases where the star is missing, indentation
is off, etc.

**Comment by billf** [comment]
lgtm

File: idl/internal/docstring.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+		// Dedent only if the first non-space character appears at or after
+		// the prefix.
+		if nonSpace >= len(prefix) {
+			lines[i] = s[len(prefix):]
+		}
+	}
+}
+
+func notIsSpace(r rune) bool {
```
Comment (kriskowal) [inline idl/internal/docstring.go]: pedantic: isNotSpace

File: idl/internal/docstring.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+// Becomes,
+//
+//  "foo"
+//  "  bar"
+//
+// If the first line was not indentend, skipFirstIfUnindented controls whether
+// we strip the indentation of the second line from all lines instead. This
+// flag is needed because our docstringcs will start at "/**" without any
+// leading whitespace.
+func dedent(lines []string, skipFirstIfUnindented bool) {
```
Comment (kriskowal) [inline idl/internal/docstring.go]: Pythonic precedent notwithstanding, pedantically, the right word is unindent. https://english.stackexchange.com/questions/149511/what-is-the-opposite-word-to-indent-outdent-or-unindent

File: idl/internal/docstring.go (hunk 7 above)
Comment (abhinav) [reply idl/internal/docstring.go]: I was going for `not(unicode.IsSpace)` but I see your point.

**Pull request merged**